rust/connlib/tunnel/src/sockets.rs

```diff
@@ -13,10 +13,41 @@ use crate::Result;
 pub struct Sockets {
     socket_v4: Option<Socket>,
     socket_v6: Option<Socket>,
+
+    #[cfg(unix)]
+    protect: Box<dyn Fn(std::os::fd::RawFd) -> io::Result<()> + Send + 'static>,
 }
 
 impl Sockets {
-    pub fn new() -> io::Result<Self> {
+    #[cfg(unix)]
+    pub fn with_protect(
+        protect: impl Fn(std::os::fd::RawFd) -> io::Result<()> + Send + 'static,
```

Comment:
@conectado As part of https://github.com/firezone/firezone/pull/4159, I want to move the initialization of the TUN device to the upper layers, i.e. outside of `Tunnel`. At that point, the use of `FIREZONE_MARK` can be entirely in the clients and the linux client can also use this callback to call `set_mark` on the socket which unifies this behaviour across all platforms.

Reply:
I'm in favor of this, in fact, I think making connlib completely platform-independent would be very nice.

Reply:
As long as it doesn't make the FFI / callbacks more complex just to do something that connlib already knows it needs to do.